File: app/src/main/java/com/atoms/purityhubserviceman/adapter/BlindAdapter.java
package com.atoms.purityhubserviceman.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.RecyclerView;


import com.atoms.purityhubserviceman.UpdateListener;

import java.util.ArrayList;
import java.util.Collections;

public abstract class BlindAdapter <T, D> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    Context mContext;
    private ArrayList<T> mArrayList;
    private UpdateListener updatelistner;
    public String valueCheck;

    public BlindAdapter(Context context, ArrayList<T> arrayList) {
        this.mContext = context;
        this.mArrayList = arrayList;
    }

    public BlindAdapter(Context mContext, ArrayList<T> mArrayList, UpdateListener updatelistner) {
        this.mContext = mContext;
        this.mArrayList = mArrayList;
        this.updatelistner = updatelistner;
    }

    public BlindAdapter(Context mContext, ArrayList<T> mArrayList, UpdateListener updatelistner, String valueCheck) {
        this.mContext = mContext;
        this.mArrayList = mArrayList;
        this.updatelistner = updatelistner;
        this.valueCheck = valueCheck;
    }

    public BlindAdapter(Context mContext, ArrayList<T> mArrayList, String valueCheck) {
        this.mContext = mContext;
        this.mArrayList = mArrayList;
        this.valueCheck = valueCheck;
    }

    public abstract int getLayoutResId();

    public abstract void onBindData(T model, int position, D dataBinding);

    protected abstract void onBindDataHolder(T t, int position, D mDataBinding, RecyclerView.ViewHolder holder);

    public abstract void onItemClick(T model, int position);

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ViewDataBinding dataBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), getLayoutResId(), parent, false);
        RecyclerView.ViewHolder holder = new ItemViewHolder(dataBinding);
        holder.setIsRecyclable(false);
        return holder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {
        onBindData(mArrayList.get(position), position, ((ItemViewHolder) holder).mDataBinding);
        onBindDataHolder(mArrayList.get(position), position, ((ItemViewHolder) holder).mDataBinding, holder);
        ((ViewDataBinding) ((ItemViewHolder) holder).mDataBinding).getRoot().setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onItemClick(mArrayList.get(position), position);
            }
        });

    }


    @Override
    public int getItemCount() {
        return mArrayList.size();
    }

    public void addItems(ArrayList<T> arrayList) {
        mArrayList = arrayList;
        this.notifyDataSetChanged();
    }

    public void swapeItem(int fromPosition, int toPosition) {
        Collections.swap(mArrayList, fromPosition, toPosition);
        this.notifyItemMoved(fromPosition, toPosition);
    }

    public void removeItem(int removeItemPosition,ArrayList<T> arrayList){
        mArrayList = arrayList;
        mArrayList.remove(removeItemPosition);
        notifyItemRemoved(removeItemPosition);
        notifyItemRangeChanged(removeItemPosition, mArrayList.size());
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    public T getItem(int position) {
        return mArrayList.get(position);
    }

    public Context getContext() {
        return mContext;
    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
        protected D mDataBinding;

        public ItemViewHolder(ViewDataBinding binding) {
            super(binding.getRoot());
            mDataBinding = (D) binding;
        }
    }

    public UpdateListener getUpdatelistner() {
        return updatelistner;
    }
}

File: app/src/main/java/com/atoms/purityhubserviceman/adapter/BlindAdapter.kt
package com.atoms.purityhubserviceman.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView
import com.atoms.purityhubserviceman.UpdateListener
import org.json.JSONArray
import java.util.*

abstract class BlindAdapter<T, D> : RecyclerView.Adapter<RecyclerView.ViewHolder> {
    var context: Context
    var arrayList: ArrayList<T>
        private set
    var updatelistner: UpdateListener? = null
        private set
    var valueCheck: String? = null
        private set
    var jsonArray: JSONArray? = null
    constructor(context: Context, arrayList: ArrayList<T>) {
        this.context = context
        this.arrayList = arrayList
    }

    constructor(mContext: Context, mArrayList: ArrayList<T>, updatelistner: UpdateListener?) {
        context = mContext
        arrayList = mArrayList
        this.updatelistner = updatelistner
    }

    constructor(
        mContext: Context,
        mArrayList: ArrayList<T>,
        updatelistner: UpdateListener?,
        valueCheck: String?
    ) {
        context = mContext
        arrayList = mArrayList
        this.updatelistner = updatelistner
        this.valueCheck = valueCheck
    }

    constructor(mContext: Context, mArrayList: ArrayList<T>, valueCheck: String?) {
        context = mContext
        arrayList = mArrayList
        this.valueCheck = valueCheck
    }

    constructor(mContext: Context, mArrayList: ArrayList<T>, jsonArray: JSONArray) {
        context = mContext
        arrayList = mArrayList
        this.jsonArray = jsonArray
    }

    abstract val layoutResId: Int
    abstract fun onBindData(model: T?, position: Int, dataBinding: D)
    protected abstract fun onBindDataHolder(
        t: T?,
        position: Int,
        mDataBinding: D,
        holder: RecyclerView.ViewHolder?
    )

    abstract fun onItemClick(model: T?, position: Int)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val dataBinding = DataBindingUtil.inflate<ViewDataBinding>(
            LayoutInflater.from(parent.context),
            layoutResId,
            parent,
            false
        )
        val holder: RecyclerView.ViewHolder = ItemViewHolder(dataBinding)
        holder.setIsRecyclable(false)
        return holder
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        onBindData(arrayList[position], position, (holder as BlindAdapter<*, *>.ItemViewHolder).mDataBinding as D)
        onBindDataHolder(arrayList[position], position, holder.mDataBinding as D, holder)
        (holder.mDataBinding as ViewDataBinding).root.setOnClickListener(View.OnClickListener {
            onItemClick(
                arrayList[position], position
            )
        })
    }

    override fun getItemCount(): Int {
        return arrayList.size
    }

    fun addItems(arrayList: ArrayList<T>) {
        this.arrayList = arrayList
        notifyDataSetChanged()
    }

    fun swapeItem(fromPosition: Int, toPosition: Int) {
        Collections.swap(arrayList, fromPosition, toPosition)
        notifyItemMoved(fromPosition, toPosition)
    }

    open fun removeItem(removeItemPosition: Int, arrayList: ArrayList<T>) {
        this.arrayList = arrayList
        arrayList.removeAt(removeItemPosition)
        notifyItemRemoved(removeItemPosition)
        notifyItemRangeChanged(removeItemPosition, arrayList.size)
    }

    open fun removeWithoutItemList(removeItemPosition: Int) {
        arrayList.removeAt(removeItemPosition)
        notifyItemRemoved(removeItemPosition)
        notifyItemRangeChanged(removeItemPosition, arrayList.size)
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    fun getItem(position: Int): T? {
        return arrayList[position]
    }

    internal inner class ItemViewHolder(binding: ViewDataBinding) :
        RecyclerView.ViewHolder(binding.root) {
        var mDataBinding: D

        init {
            mDataBinding = binding as D
        }
    }
}